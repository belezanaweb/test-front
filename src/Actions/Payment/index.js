export const saveProduct = (data) => {
  return {
    type: '@reducer/ADD_PRODUCTS',
    payload: data
  }
}

export const saveTotal = (data) => {
  return {
    type: '@reducer/ADD_TOTALS',
    payload: data
  }
}
