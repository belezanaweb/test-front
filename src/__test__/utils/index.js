import checkPropTypes from 'check-prop-types'

export const findByTestAtrr = (component, attr) => {
  const wrapper = component.find(`[data-test='${attr}']`)
  return wrapper
}

export const checkProps = (component, expectedProps) => {
  const propsError = checkPropTypes(component.propTypes, expectedProps, 'props', component.name)
  return propsError
}
