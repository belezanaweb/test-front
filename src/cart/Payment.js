import React, { Component } from 'react';
import { connect } from 'react-redux';
import * as actionCreators from '../store/actions/index.js';
import FinalizeAccount from '../components/FinalizeAccount';
import Header from '../components/Header';
import Button from '../components/Button';
import UserForm from '../components/UserForm'

class Payment extends Component {
  render() {
    let isDisabled =
      !this.props.cardName ||
      !this.props.cardCvv ||
      !this.props.cardDate ||
      !this.props.cardNumber;

    return (
      <div>
        <Header />
        <UserForm 
          cardName={this.props.cardName} 
          cardCvv={this.props.cardCvv} 
          cardDate={this.props.cardDate} 
          cardNumber={this.props.cardNumber} 
        />
        <FinalizeAccount
          subTotal={this.props.subTotal}
          shippingTotal={this.props.shippingTotal}
          discount={this.props.discount}
          total={this.props.total}
        />
        <Button
          label='FINALIZAR O PEDIDO'
          handleClick='success'
          disabled={isDisabled}
        />
      </div>
    );
  }

}
const mapStateToProps = (state) => {
  return state
};

export default connect(mapStateToProps, actionCreators)(Payment);
