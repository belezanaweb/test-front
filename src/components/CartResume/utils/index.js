export const normalizeCartData = data => {
  return {
    productsQuantity: data.items.length,
    subTotal: data.subTotal.toFixed(2).replace('.', ','),
    shipping: data.shippingTotal.toFixed(2).replace('.', ','),
    discount: data.discount.toFixed(2).replace('.', ','),
    total: data.total.toFixed(2).replace('.', ','),
  };
};
