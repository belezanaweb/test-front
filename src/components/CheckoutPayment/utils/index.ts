import * as Yup from 'yup'

const errors = {
  name: 'Insira um nome válido',
  code: 'Insira um código válido',
  number: 'Insira um número válido',
  expiration: 'Insira uma data válida',
}

export const initialValues = {
  creditCardName: '',
  creditCardNumber: '',
  creditCardExpiration: '',
  creditCardCode: '',
}

const regexOnlyNumbers = /[^\d.-]+/g

export const CheckoutPaymentSchema = Yup.object().shape({
  creditCardName: Yup.string()
    .min(2, errors.name)
    .matches(/^([A-Za-z\u00C0-\u00D6\u00D8-\u00f6\u00f8-\u00ff\s]*)$/gi, errors.name)
    .required(errors.name),
  creditCardNumber: Yup.string()
    .min(2, errors.number)
    .test('len', errors.number, (val) => {
      const value = val.replace(regexOnlyNumbers, '')

      return value.length === 16
    })
    .required(errors.number),
  creditCardExpiration: Yup.string()
    .min(2, errors.name)
    .test('len', errors.expiration, (val) => {
      const value = val.replace(regexOnlyNumbers, '')

      return value.length === 4
    })
    .required(errors.expiration),
  creditCardCode: Yup.string()
    .matches(/^[0-9]+$/, errors.code)
    .test('len', errors.code, (val) => val.length === 3)
    .required(errors.code),
})
