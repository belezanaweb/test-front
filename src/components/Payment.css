.payment {
  max-width: 640px;
  margin: 0 auto;
  padding: 0 10px;
}

.payment .credit-card {
  background: white;
  box-shadow: 1px 1px 5px 0 rgba(0,0,29,0.22);
  border-radius: 3px;
  color: #ccc;
  padding: 10px;
}

.payment .credit-card label {
  margin-bottom: 5px;
  display: inline-block;
  font-size: 12px;
  font-weight: bold;
  line-height: 14px;
}

.payment .credit-card .field {
  padding-right: 20px;
  position: relative;
}

.payment .credit-card .field {
  margin-bottom: 20px
}

.payment input {
  display: block;
  width: 100%;
  border: 1px solid #E7E7E7;
  box-shadow: inset 0 1px 2px 0 rgba(0,0,0,0.2);
  border-radius: 3px;
  padding: 14px 0 12px 13px
}

.payment input:focus {
  border-color: #A43287;
}

.payment .invalid {
  border-color: #F30;
}

.payment .error {
  position: absolute;
  color: #F30;
  font-size: 12px;
  bottom: -17px;
}
