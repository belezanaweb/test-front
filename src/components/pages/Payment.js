import React, {useState, useEffect } from 'react';
import UIButton from '../commons/UIButton';
import Grid from '@material-ui/core/Grid';
import { makeStyles } from '@material-ui/core/styles';
import UIHeader from '../commons/UIHeader';
import UIPaper from '../commons/UIPaper';
import UIPurchaseSummary from '../commons/UIPurchaseSummary';
import UIInput from '../commons/UIInput';
import CardValidator from 'card-validator';
import { useHistory } from "react-router-dom";
import moment from 'moment';
import { connect } from 'react-redux';
import userModel from '../../model/userModel';

const useStyles = makeStyles(theme=>({
    root: {
        color: '#CCC',
        padding: theme.spacing(0),
        width: '100%'
    },
}));

function Payment( props ){
    const classes = useStyles();
    const history = useHistory();
    const [validCreditcard, setValidCreditcard] = useState(false);
    const [validDate, setValidDate] = useState(false);
    const [validCVV, setValidCVV] = useState(false);
    const [submit, setSubmit] = useState(false);
    const [values, setValues] = useState( props.user || {
        creditcard: '',
        name: '',
        validate: '',
        cvv: ''
    });

    const isValidCreditcard = (e) => {
        setValidCreditcard(!CardValidator.number(e.target.value.replace(/\./g,'')).isValid);
    }

    const isValidDate = (e) => {
        let year = e.target.value.slice(-4).replace(/\//g, '');
        let month = e.target.value.slice(0,2).replace(/\//g, '');
        let valid = false;

        if(
            month.match(/[0-9]{2}/gi) &&
            year.match(/[0-9]{4}/gi) &&
            parseInt(year) >= moment().year() && 
            parseInt(year) <= (moment().year() + 30)
        )
            valid = moment(`${year}-${month}-01`).isValid();
        else
            valid = false;
        
        setValidDate(!valid);
    }

    const isValidCVV = (e) => {
        setValidCVV( e.target.value.length < 3  );
    }

    const doSubmit = (e) => {
        e.preventDefault();
        props.dispatch(userModel.setUser(values));
        history.push('/confirmation');
    }

    const handleChange = ( evt ) => {
        evt.persist();
        setValues({...values, [evt.target.name]: evt.target.value});
    }

    useEffect(()=>{
        if(!props.products.length){
            history.push('/cart');
        }
    }, [props.products.length]);

    useEffect(()=>{
        setSubmit(!(!validCreditcard && !validDate && !validCVV && Object.values(values).indexOf('') === -1));
    }, [values]);

    return (
        <React.Fragment>
            <UIHeader step={1} />
            <Grid container className={classes.root}>
                <Grid item xs={12}>
                    <form onSubmit={doSubmit}>
                        <UIPaper title="CARTÃO DE CRÉDITO">
                            <Grid container spacing={1}>
                                <Grid item xs={12}>
                                    <UIInput 
                                        type="creditcard" 
                                        name="creditcard"
                                        required 
                                        onChange={(e)=>{
                                            handleChange(e);
                                            isValidCreditcard(e);
                                        }}
                                        label="Número do Cartão" 
                                        placeholder="____.____.____.____.____" 
                                        error={validCreditcard}  
                                        infoempty="Obrigatório informar um cartão de crédito válido"
                                        infoerror="O número do cartão de crédito é inválido"
                                        value={values.creditcard}
                                    />
                                </Grid>
                                <Grid item xs={12}>
                                    <UIInput 
                                        label="Nome do titular"
                                        name="name"
                                        required placeholder="Como no cartão"
                                        value={values.name}
                                        onChange={(e)=>{
                                            handleChange(e);
                                        }}
                                    />
                                </Grid>

                                <Grid item xs={6}>
                                    <UIInput
                                        label="Validade (mês/ano)" 
                                        name="validate" 
                                        required type="date" 
                                        onChange={(e)=>{
                                            handleChange(e);
                                            isValidDate(e);
                                        }}
                                        error={validDate}
                                        infoerror="Digite uma data válida" 
                                        infoempty="A data de validade deve ser informada" 
                                        placeholder="__/____" 
                                        value={values.validate}
                                    />
                                </Grid>
                                <Grid item xs={6}>
                                    <UIInput label="CVV" 
                                        mask={[/\d/, /\d/, /\d/]} 
                                        required 
                                        onChange={(e)=>{
                                            handleChange(e);
                                            isValidCVV(e);
                                        }} 
                                        name="cvv"
                                        error={validCVV} 
                                        infoerror="O valor deve conter 3 dígitos" 
                                        infoempty="O campo CVV deve ser preenchido" 
                                        maxLength="3" 
                                        placeholder="___" 
                                        value={values.cvv}
                                    />
                                </Grid>
                            </Grid>
                        </UIPaper>
                    </form>
                </Grid>
                <Grid item xs={12}>
                    <UIPurchaseSummary summary={props.summary}></UIPurchaseSummary>
                </Grid>
                <Grid item xs={12}>
                    <UIButton disabled={ submit } onClick={doSubmit}>
                        Finalizar o pedido
                    </UIButton>
                </Grid>
            </Grid>
        </React.Fragment>
    )
}

export default connect(store => ({ ...store }))(Payment);