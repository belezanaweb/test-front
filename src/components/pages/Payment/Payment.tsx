import { useContext, useEffect } from 'react'
import { useNavigate } from 'react-router-dom'
import { CartContext } from '../../../contexts/CartContext'
import { HeaderContext } from '../../../contexts/HeaderContext'
import Button from '../../atoms/Button/Button'
import { Title } from '../../atoms/Title/Title.styles'
import ProductPrices from '../../molecules/ProductPrices/ProductPrices'
import * as Style from './Payment.styles'

const Payment: React.FC = (props) => {
  const navigate = useNavigate()
  const { setActiveItem } = useContext(HeaderContext)
  const { products } = useContext(CartContext)

  const handleNextStep = () => {
    navigate('/confirmation')
  }

  useEffect(() => {
    setActiveItem(2)
    if (!products) {
      navigate('/')
    }
  }, [products, navigate, setActiveItem])

  return (
    <Style.Container>
      <Title>cartão de crédito</Title>
      {products && <ProductPrices productCart={products} />}
      <Button type="button" text="finalizar pedido" onClick={handleNextStep} />
    </Style.Container>
  )
}

export default Payment
