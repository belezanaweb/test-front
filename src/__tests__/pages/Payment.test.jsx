import React from 'react'
import { render } from '../../utils/test-utils'
import Payment from '../../pages/Payment'

jest.mock('../../hooks/checkout.jsx', () => ({
  useCheckout: jest.fn().mockReturnValue({
    cartItems: [],
    cartSummary: {},
    isLoading: false
  })
}))

describe('Payment Page', () => {
  test('should render without errors', () => {
    const { container } = render(<Payment />)
    expect(container).toBeInTheDocument()
  })

  test('should match with snapshot', () => {
    const { container } = render(<Payment />)
    expect(container).toMatchSnapshot()
  })
})
