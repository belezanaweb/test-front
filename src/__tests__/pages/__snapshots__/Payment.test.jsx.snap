// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Payment Page should match with snapshot 1`] = `
<div>
  <header
    class="sc-gtsrHT knapBr"
  >
    <div
      class="sc-dlnjwi fieBIy"
    >
      <nav
        class="sc-hKFxyN gbRFWY"
      >
        <a
          class="sc-eCApnc jdlbal"
          href="/cart"
        >
          Sacola
        </a>
        <a
          class="sc-eCApnc jdlbal"
          href="/payment"
        >
          Pagamento
        </a>
        <a
          class="sc-eCApnc jdlbal"
          href="/confirmation"
        >
          Confirmação
        </a>
      </nav>
    </div>
  </header>
  <main
    class="sc-bYwzuL kIRxIx"
  >
    <div
      class="sc-kLojOw hHvdtz"
    >
      <div
        class="sc-iklJeh kANenp"
      >
        <section
          class="sc-jSFjdj"
        >
          <h2
            class="sc-gKAaRy iuHTKJ"
          >
            Cartão de Crédito
          </h2>
          <div
            class="sc-iCoGMd oVyrW"
          >
            <form
              class="sc-lmgQwP cgxYwA"
            >
              <div
                class="sc-bCwfaz cOSPOl"
              >
                <label
                  class="sc-iwajpm khNqxB"
                >
                  Número do cartão
                </label>
                <input
                  class="sc-fKgJPI ArQqe"
                  placeholder="____.____.____.____"
                  value=""
                />
              </div>
              <div
                class="sc-bCwfaz cOSPOl"
              >
                <label
                  class="sc-iwajpm khNqxB"
                >
                  Nome do Titular
                </label>
                <input
                  class="sc-kfYoZR cXtyio"
                  placeholder="Como no cartão"
                  value=""
                />
              </div>
              <div
                class="sc-iJCRrE dXXKOn"
              >
                <div
                  class="sc-bCwfaz cOSPOl"
                >
                  <label
                    class="sc-iwajpm khNqxB"
                  >
                    Validade (mês/ano)
                  </label>
                  <input
                    class="sc-fKgJPI ArQqe"
                    placeholder="__/____"
                    value=""
                  />
                </div>
                <div
                  class="sc-bCwfaz cOSPOl"
                >
                  <label
                    class="sc-iwajpm khNqxB"
                  >
                    CVV
                  </label>
                  <input
                    class="sc-fKgJPI ArQqe"
                    placeholder="____"
                    value=""
                  />
                </div>
              </div>
            </form>
          </div>
        </section>
        <section
          class="sc-dIsUp gMJrvn"
        >
          <div
            class="sc-hBMUJo sc-fnVZcZ jcqzAV"
          >
            <p
              class="sc-bqGGPW gxsOhh"
            >
              Produtos
            </p>
            <p
              class="sc-ksluID gMSQZp"
            >
              R$ NaN
            </p>
          </div>
          <div
            class="sc-hBMUJo sc-fFSPTT jcqzAV"
          >
            <p
              class="sc-bqGGPW gxsOhh"
            >
              Frete
            </p>
            <p
              class="sc-ksluID gMSQZp"
            >
              R$ NaN
            </p>
          </div>
          <div
            class="sc-hBMUJo sc-bkbkJK jcqzAV gtxaIs"
          >
            <p
              class="sc-bqGGPW gxsOhh"
            >
              Desconto
            </p>
            <p
              class="sc-ksluID gMSQZp"
            >
              R$ NaN
            </p>
          </div>
          <div
            class="sc-hBMUJo sc-iemWCZ jcqzAV fJDoAN"
          >
            <p
              class="sc-bqGGPW gxsOhh"
            >
              Total
            </p>
            <p
              class="sc-ksluID gMSQZp"
            >
              R$ NaN
            </p>
          </div>
        </section>
      </div>
      <button
        class="sc-bdnxRM grNhhY"
      >
        Finalizar o Pedido
      </button>
    </div>
  </main>
</div>
`;
